File: 22-03-2025_Count_the_Number_of_Complete_Components.java
class Solution {
    public int countCompleteComponents(int n, int[][] edges) {

       List<List<Integer>> adjList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adjList.add(new ArrayList<>());
        }

        for (int[] edge : edges) {
            int v1 = edge[0];
            int v2 = edge[1];
            adjList.get(v1).add(v2);
            adjList.get(v2).add(v1);
        }

        boolean[] visited = new boolean[n];
        int ans = 0;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                Set<Integer> component = new HashSet<>();
                DFS(i, adjList, visited, component);

                boolean complete = true;
                int nodeCount = component.size();

                for (int v : component) {
                    if (adjList.get(v).size() != nodeCount - 1) {
                        complete = false;
                        break;
                    }
                }

                if (complete) {
                    ans++;
                }
            }
        }

        return ans;
    }

    private void DFS(int s, List<List<Integer>> adjList, boolean[] visited, Set<Integer> component) {
        if (!visited[s]) {
            visited[s] = true;
            component.add(s);
        }

        for (int v : adjList.get(s)) {
            if (!visited[v]) {
                DFS(v, adjList, visited, component);
            }
        }
    }
}

File: 22-03-2025_Count_the_Number_of_Complete_Components.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        adjList = [[] for _ in range(n)]

        for v1, v2 in edges:
            adjList[v1].append(v2)
            adjList[v2].append(v1)

        visited = [False] * n

        def dfs(node: int, component: Set[int]):
            if not visited[node]:
                visited[node] = True
                component.add(node)
                for neighbor in adjList[node]:
                    if not visited[neighbor]:
                        dfs(neighbor, component)

        ans = 0

        for i in range(n):
            if not visited[i]:
                component = set()
                dfs(i, component)

                node_count = len(component)
                complete = all(len(adjList[v]) == node_count - 1 for v in component)

                if complete:
                    ans += 1

        return ans
